[FX] Support wrapping builtins with wrap

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50184 [FX] Support wrapping builtins with wrap**

- Do some trickery to detect if we're patching a builtin that hasn't been overridden in the global frame of the wrap() caller. In this case, patch the builtin function into the global namespace but add a None sentinel to indicate we should delete the function after we're done.
- Make wrap() more exception safe

Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

Commit:
[FX] Make len traceable and scriptable with wrap

[ghstack-poisoned]

Commit:
Update on "[FX] Make len traceable and scriptable with wrap"


Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]

Commit:
Update on "[FX] Make len traceable and scriptable with wrap"


Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]

Commit:
Update on "[FX] Make len traceable and scriptable with wrap"


Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]

Commit:
Update on "[FX] Make len traceable and scriptable with wrap"


Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]

Commit:
Update on "[FX] Make len traceable and scriptable with wrap"


Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]

Commit:
Update on "[FX] Make len traceable and scriptable with wrap"


Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]

zdevito commented on test/test_fx.py:
```diff
@@ -51,6 +51,11 @@ def a_lifted_leaf2(a, b):
+def fx_len(item):
+    return len(item)
+
+wrap(fx_len)
```
Shouldn't `wrap(len)` work?

jamesr66a replied:
Builtins don't have `__code__`:

```
Traceback (most recent call last):
  File "test/test_fx.py", line 58, in <module>
    wrap(len)
  File "/data/users/jamesreed/pytorch/torch/fx/symbolic_trace.py", line 415, in wrap
    fn_name = fn_or_name.__code__.co_name
AttributeError: 'builtin_function_or_method' object has no attribute '__code__'
```

Commit:
Update on "[FX] Test that len is symtraceable and scriptable with wrap()"


Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]

Commit:
Update on "[FX] Support wrapping builtins with wrap"


- Do some trickery to detect if we're patching a builtin that hasn't been overridden in the global frame of the wrap() caller. In this case, patch the builtin function into the global namespace but add a None sentinel to indicate we should delete the function after we're done.
- Make wrap() more exception safe

Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]

Commit:
Update on "[FX] Support wrapping builtins with wrap"


- Do some trickery to detect if we're patching a builtin that hasn't been overridden in the global frame of the wrap() caller. In this case, patch the builtin function into the global namespace but add a None sentinel to indicate we should delete the function after we're done.
- Make wrap() more exception safe

Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]

Commit:
Update on "[FX] Support wrapping builtins with wrap"


- Do some trickery to detect if we're patching a builtin that hasn't been overridden in the global frame of the wrap() caller. In this case, patch the builtin function into the global namespace but add a None sentinel to indicate we should delete the function after we're done.
- Make wrap() more exception safe

Differential Revision: [D25819832](https://our.internmc.facebook.com/intern/diff/D25819832)

[ghstack-poisoned]
